core/src/main/java/ai/timefold/solver/core/impl/score/stream/collector/connected_ranges/ConnectedRangeImpl.java

```diff
@@ -0,0 +1,212 @@
+package ai.timefold.solver.core.impl.score.stream.collector.connected_ranges;
+
+import java.util.Iterator;
+import java.util.NavigableSet;
+import java.util.Objects;
+import java.util.function.BiFunction;
+
+import ai.timefold.solver.core.api.score.stream.common.ConnectedRange;
+
+final class ConnectedRangeImpl<Interval_, Point_ extends Comparable<Point_>, Difference_ extends Comparable<Difference_>>
+        implements ConnectedRange<Interval_, Point_, Difference_> {
+
+    private final NavigableSet<IntervalSplitPoint<Interval_, Point_>> splitPointSet;
+    private final BiFunction<? super Point_, ? super Point_, ? extends Difference_> differenceFunction;
+    private IntervalSplitPoint<Interval_, Point_> startSplitPoint;
+    private IntervalSplitPoint<Interval_, Point_> endSplitPoint;
+
+    private int count;
+    private int minimumOverlap;
+    private int maximumOverlap;
+    private boolean hasOverlap;
+
+    ConnectedRangeImpl(NavigableSet<IntervalSplitPoint<Interval_, Point_>> splitPointSet,
+            BiFunction<? super Point_, ? super Point_, ? extends Difference_> differenceFunction,
+            IntervalSplitPoint<Interval_, Point_> start) {
+        if (start == null) {
+            throw new IllegalArgumentException("start (" + start + ") is null");
+        }
+        if (differenceFunction == null) {
+            throw new IllegalArgumentException("differenceFunction (" + differenceFunction + ") is null");
+        }
```

Comment:
For the sake of bravity, I'd prefer

    this.startSplitPoint = Objects.requireNonNull(start, "start");